### PR TITLE
Safer code in NativeUISystem::DisplayBlockingDialog (iOS)

### PR DESCRIPTION
**Pull request by alexeykostin**
## What does this PR do?

In case a deadlock while attempting to display `NativeUISystem::DisplayBlockingDialog` (iOS) is detected, save the dialog data in a file and abort the app

## How was this PR tested?

Tested using a local build, iPhone 15.

### CODE REVIEW

File: Code/Framework/AzCore/Platform/iOS/AzCore/NativeUI/NativeUISystemComponent_iOS.mm [hunk 1]
```diff
@@ -10,16 +10,65 @@
...
+            
+            // Include the date in the file name to not overwrite previous artifacts.
+            NSString* fileName = [NSString stringWithFormat:@"BlockingDialogAborted %@.txt", dateString];
+            
+            NSArray* allPaths = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES);
+            NSString* documentsDirectory = [allPaths objectAtIndex:0];
+            NSString* pathForLog = [documentsDirectory stringByAppendingPathComponent:fileName];
+
+            // Redirect NSLog messages to the file.
+            freopen([pathForLog cStringUsingEncoding:NSASCIIStringEncoding], "a+", stderr);
```
Comment (aefimov3) [inline Code/Framework/AzCore/Platform/iOS/AzCore/NativeUI/NativeUISystemComponent_iOS.mm:58]: Interesting that there is stderr, it looks like normal NSLog goes to stderr. I hope you checked it,

File: Code/Framework/AzCore/Platform/iOS/AzCore/NativeUI/NativeUISystemComponent_iOS.mm [hunk 2]
```diff
@@ -10,16 +10,65 @@
...
+            
+            // These messages go to the file.
+            NSLog(@"BlockingDialog failed to display at %@", dateString);
+            
+            NSLog(@"BlockingDialog title is %@", nsTitle);
```
Comment (aefimov3) [inline Code/Framework/AzCore/Platform/iOS/AzCore/NativeUI/NativeUISystemComponent_iOS.mm]: I believe just the title and the message is OK without extra text.

File: Code/Framework/AzCore/Platform/iOS/AzCore/NativeUI/NativeUISystemComponent_iOS.mm [hunk 3]
```diff
@@ -75,8 +130,19 @@
                     }
                     dispatch_semaphore_signal(blockSem);
                 });
+                
                 // Wait till the user responds to the message box.
-                dispatch_semaphore_wait(blockSem, DISPATCH_TIME_FOREVER);
+                static constexpr int WaitMainThreadMs = 10000;
```
Comment (aefimov3) [inline Code/Framework/AzCore/Platform/iOS/AzCore/NativeUI/NativeUISystemComponent_iOS.mm]: That's too long, the app can be terminated by user, please reduce to 3 seconds.

File: Code/Framework/AzCore/Platform/iOS/AzCore/NativeUI/NativeUISystemComponent_iOS.mm (hunk 2 above)
Comment (alexeykostin) [reply Code/Framework/AzCore/Platform/iOS/AzCore/NativeUI/NativeUISystemComponent_iOS.mm]: Ok

File: Code/Framework/AzCore/Platform/iOS/AzCore/NativeUI/NativeUISystemComponent_iOS.mm (hunk 3 above)
Comment (alexeykostin) [reply Code/Framework/AzCore/Platform/iOS/AzCore/NativeUI/NativeUISystemComponent_iOS.mm]: Reduced

**Review by aefimov3 — APPROVED** [review]

**Review by Twolewis — APPROVED** [review]

**Pull request merged**